### ISSUE TITLE
Completed work

### ISSUE THREAD

**Issue by joao-pm-santos96** [lardemua/SmObEx#6]
# Week 1

- First time retrieving and visualizing single view point cloud and octomap

![rviz_screenshot_2019_03_06-17_14_02](https://user-images.githubusercontent.com/47630528/54637199-f7c9a480-4a7f-11e9-92c0-53b1f6e7700c.png)
![rviz_screenshot_2019_03_06-17_15_10](https://user-images.githubusercontent.com/47630528/54637200-f7c9a480-4a7f-11e9-9967-534881c95da9.png)

**Comment by joao-pm-santos96** [comment]
# Week 2

- (kinda) installed ROS-I on manipulator 

- Created FANUC M-6iB/6S support and moveit_configure packages
   
![fanuc_m6ib6s_implement](https://user-images.githubusercontent.com/47630528/54637266-23e52580-4a80-11e9-9024-e516a6e08346.png)

- Connected ROS to the FANUC

- Calibrated intrinsic and extrinsic camera parameters 

![calib_rviz](https://user-images.githubusercontent.com/47630528/54637343-598a0e80-4a80-11e9-93f4-f3a94bf58429.png)

(video is os ReadME)

**Comment by joao-pm-santos96** [comment]
## march 19 

- retrived multi view point cloud and octomap 

![LAR_360_octomap](https://user-images.githubusercontent.com/47630528/54637440-8dfdca80-4a80-11e9-9584-d732f5c23b84.png)

![LAR_360_pointCloud](https://user-images.githubusercontent.com/47630528/54637441-8dfdca80-4a80-11e9-8c11-a5110ceb265b.png)

**Comment by joao-pm-santos96** [comment]
Hi @miguelriemoliveira and @rarrais, another update. Has able to restrict the volume to create the OctoMap. Here's an example:

![inside_octomap](https://user-images.githubusercontent.com/47630528/54698652-8cd3a880-4b27-11e9-93d1-fca5d95c5df4.png)

**Comment by miguelriemoliveira** [comment]
Great!

**Comment by miguelriemoliveira** [comment]
Hi @joao-pm-santos96 , unless you need @miguelriemoliveira or @rarrais to do something on this topic,
we should not be assignees. 

Instead, just mention us at some point during the issue.

I usually do like this:

FYI (for your information), @rarrais  and @miguelriemoliveira 

That way we always receive emails with any updates.

I will remove me and @rarrais from assignees.

**Comment by joao-pm-santos96** [comment]
Hi @miguelriemoliveira I did that because I thought that by being assigned I wouldn't need to mention neither of you, and both would still received e-mail notification. Thanks for the advice!

**Comment by joao-pm-santos96** [comment]
Hello @miguelriemoliveira, @rarrais just to let you know the situation point.

This video (https://youtu.be/pa0htI7LZPg) shows the current state of the work. The volume to explore is now restricted. Only thing remaining it to make sure the timestamps are correct between the robots' tf and the OctoMap map.

**Comment by miguelriemoliveira** [comment]
Hi @joao-pm-santos96, if you assign me an issue that means I have to do something, so I get stressed :)

**Comment by joao-pm-santos96** [comment]
# march 22
@miguelriemoliveira @rarrais 
Maybe the last update of the week:

- Proved that MoveIt is working by connecting to Roboguide in a separate windows machine (found some issues, they are now solved I hope)
- Continue to use point_cloud_spatial_filter to define the volume, but to define that volume (to have higher hz in "operation mode") used pcl nodelets instead.

Will now try to make a good video about the reconstruction and when done I'll publish.

Have a good weekend

**Comment by miguelriemoliveira** [comment]
Thanks for the update!

*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by joao-pm-santos96** [comment]
Hello @miguelriemoliveira and @rarrais, created a simple node to generate a given number of poses inside a volume and publish them. By default it generates 20 and all inside the defined volume, but that can be change with params.

![Screenshot from 2019-03-26 16-29-55](https://user-images.githubusercontent.com/47630528/55015288-783e5700-4fe4-11e9-9742-1005b9bb962b.png)

**Comment by joao-pm-santos96** [comment]
Well it's been a while since I updated this issue, my apologies @miguelriemoliveira and @rarrais. I'm currently working on improving the performance of finding the unknown space [more info here](https://github.com/miguelriemoliveira/octomap_tools/issues/3).  

Relatively to the pose generator, it is done since Tuesday and now accepts a 3D point to look at and the min and max radius. In the screenshot, te point is (0,0,0) and the radius are 0.8 and 1.2 respectively. What the node does is calling a function which generates one pose (looking to the point that is a function's parameter and yy axis tending to point down) as many times as wanted.

![Screenshot from 2019-04-06 16-17-15](https://user-images.githubusercontent.com/47630528/55671454-0db0d500-5888-11e9-98f6-099e7254fd2f.png)

**Comment by joao-pm-santos96** [comment]
Hi @miguelriemoliveira @rarrais ! As a fisrt small step I'm now able to visualize the ray projected from each pose (for now it's just one, but the implementation to show all the rays projected from the camera is just a matter of a for cycle) and count the unknown cells intercepted by that ray until a occupied voxel or the bounds of the octomap are reached (not what we fully want but fix should not be hard).  

![Screenshot from 2019-04-10 10-20-30](https://user-images.githubusercontent.com/47630528/55867465-70e98280-5b7a-11e9-9ec7-3e75e23e4822.png)

**Comment by joao-pm-santos96** [comment]
A better view (two of the poses are not right, must fix there orientation):

![Screenshot from 2019-04-10 10-26-35](https://user-images.githubusercontent.com/47630528/55867868-39c7a100-5b7b-11e9-8ca9-0a1081518bc6.png)

**Comment by joao-pm-santos96** [comment]
Hello @miguelriemoliveira and @rarrais ! I'll try to recap this week work:

1. got an away to evaluate each pose, by counting only once all the voxels that are passed by all the poses' rays.
2. if a ray intercepts some unknowns and an ocuppied inside the 0.8m FOV, this ones are not taken in account 
3. marked as blue the space that will potentially by discovered by that pose (figure)

It's all if I remember correctly, doesn't seam much but it was hard work.

See you Monday!

![Screenshot from 2019-04-12 17-37-24](https://user-images.githubusercontent.com/47630528/56052736-1f8fed80-5d4a-11e9-8eeb-67bececd6b43.png)

**Comment by miguelriemoliveira** [comment]
Hi @joao-pm-santos96 ,

The image looks great. And I think it looks like hard work and a lot ... this week you were creating new stuff.

Next week lets work for some hours toghether to further advance

**Comment by joao-pm-santos96** [comment]
Made a slight update to the color scheme, now the blue ones are the first to be intersected by some ray and the light blue ones are the remaining. Not sure if confusing or not, seamed a good ideia to me, since they will have different weights in the score...

Like that idea a lot! Would be a great help!

![Screenshot from 2019-04-13 23-08-59](https://user-images.githubusercontent.com/47630528/56085790-4a10a200-5e41-11e9-90b8-4e07177855d0.png)

**Comment by joao-pm-santos96** [comment]
Hi @miguelriemoliveira 

like we talked yesterday, it's now possible to move the robot with an interactive marker and evaluate that position.

![Screenshot from 2019-04-16 12-12-14](https://user-images.githubusercontent.com/47630528/56205233-1603cf80-6041-11e9-8a58-d8faf75ae914.png)

**Comment by joao-pm-santos96** [comment]
On the afternoon will check by actually moving the robot

**Comment by miguelriemoliveira** [comment]
Great news!

**Comment by joao-pm-santos96** [comment]
Hi @miguelriemoliveira and @rarrais , I'm sending you a video o the latest updates to the work: https://youtu.be/ltMPFWkhAAE .

The manual exploration in mostly done now. Thanks!

**Comment by miguelriemoliveira** [comment]
Hi,

the video looks great.

https://youtu.be/ltMPFWkhAAE

I do have some suggestions to improve it but the overall quality is very
good.
Congratulations João!

Suggestions:

1. 1m40, the robot does not move and you appear in the right bottom video
...
2. Captions should be inserted with youtube video editor for easy update
(are they?).
3. When starting the video, show the point cloud of the scene, and explain
that the robot cannot see the whole scene with a single view.
4. Then show the octomap and explain occupied and free voxels, and then the
unknown space (explain the colors).
5. Explain how the evaluation works by showing the camera fustrum (not yet
well drawn) as well as the rays used for assessing which unknown voxels
will (may ) become known
6. Explain that we use moveit to condition the possible poses to poses
which are reachable by the manipulator
7. Show with more detail how the blue voxels (unknown space, to be known
after moving to the evaluated pose) disapear once the robot moves to that
pose and the octomap is updated.
8. I have mixed feelings about putting the video in fast forward. I don't
like it and, moreover, most nowdays video players (youtube for example)
allow the viewer to set the playing to 2x or any ratio he wants. Thus, it
is pointless to accelerate the video, the human viewer may do it if he
wants to.

Sorry for all the suggestions, got carried away by the quality of the video.

I am sending the video to Prof Vitor Santos and Prof Paulo Dias (in cc) as
well, since we discussed this during lunch.

Regards,
Miguel
*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by miguelriemoliveira** [comment]
Hi João,

one more comment: you should add to rviz the rgb image of the hand held
camera, that should be nice to see.

Regards,
Miguel

*  Miguel Riem de Oliveira                                             *
email:    mriem@ua.pt    |    m.riem.oliveira@gmail.com
Dep. de Engenharia Mecânica, Universidade de Aveiro
*Campus Universitário de Santiago, 3810-193  Portugal*

**Comment by joao-pm-santos96** [comment]
Hi @miguelriemoliveira !

Really appreciate the suggestions, and it's a good thing that they will all be saved in this issue for me to recap them latter when doing a more "final" video.

Thnaks!

**Comment by joao-pm-santos96** [comment]
Hello @miguelriemoliveira and @rarrais 

At this stage, the package generates poses mostly reachable by the manipulator and chooses the best one. For each cluster are generated n poses pointing towards its center.

So I did de profiling (with step 20) of the code and this is what I've found:

Pose generation: ~0.07s
Planning: ~0.19s
Evaluating pose: ~0.25s (for 768 rays) 

So the two big constrains are planning and evaluation. In evaluation, each ray takes about 0.0003s which is not that much, it's just the number of rays that is big (and would be bigger with step 1...). About planning, maybe there are faster algorithms, its a matter of investigate.

![Screenshot from 2019-04-24 21-58-39](https://user-images.githubusercontent.com/47630528/56693362-291b3d00-66dc-11e9-9c52-4c98f51f5284.png)
